src/arch/mips/lib/debug/stack_iter.c

```diff
@@ -0,0 +1,86 @@
+/**
+ * @brief
+ * @author Nastya Nizharadze
+ * @date 30.07.2019
+ */
+
+#include <stddef.h>
+#include <stdint.h>
+#include <hal/context.h>
+#include <util/lang.h>
+#include <util/math.h>
+#include <assert.h>
+#include "stack_iter.h"
+#include "debug/symbol.h"
+
+/* Mips stack:
+ *
+ * high address ^
+ * 
+ *  -------- frame begining
+ * |        | ---		<- new fp + offset
+ * |   ra   |  ^		<- new fp + offset - 4
+ * | old fp |  |		<- new fp + offset - 8
+ * |   ...  |  | offset
+ * |        |  |
+ * |        |  v
+ * |        | ---		<- new fp
+ *  -------- frame ending
+ * 
+ *  */
+
+static int get_offset(void *ra) {
+	uint32_t opcode;
+	const struct symbol *a;
+
+/* Find an address of first function's instruction */
+	a = symbol_lookup(ra);
+	if (a == NULL) {
+		return 0;
+	}
+/* Find an operation code. it is always addiu sp,sp,x 
+ * where x is an offset */
+	opcode = *((unsigned *) (a->addr));
+/* The operation code for addiu sp,sp,x should always started whith 0X27bd */
+	assert((opcode & 0xffff0000) == 0x27bd0000);
+/* The last 4 digits are for offset*/
+	return (0x10000 - (opcode & 0xffff));
```

Comment:
I'd propose to specify directly that this function assumes offset _must_ be negative.

```
int16_t offset;
offset = opcode & 0xffff;
assert(offset < 0);
return -offset;
```

Reply:
done